docs/RFC/draft-paragonie-paseto-rfc-00.txt

```diff
@@ -0,0 +1,1680 @@
+    | exp | Expiration | DtTime | {"exp":"2039-01-01T00:00:00+00:00"} |
+    | nbf | Not Before | DtTime | {"nbf":"2038-04-01T00:00:00+00:00"} |
+    | iat | Issued At  | DtTime | {"iat":"2038-03-17T00:00:00+00:00"} |
+    | jti |  Token ID  | string |  {"jti":"87IFSGFgPNtQNNuw0AtuLttP"} |
+    +-----+------------+--------+-------------------------------------+
+
+   In the table above, DtTime means an ISO 8601 compliant DateTime
+   string.
+
+   Any other claims can be freely used.  These keys are only reserved in
+   the top-level JSON object.
+
+   The keys in the above table are case-sensitive.
+
+   Implementors SHOULD provide some means to discourage setting invalid/
+   arbitrary data to these reserved claims.
+
+6.1.1.  Key-ID Support
+
+   Some systems need to support key rotation, but since the payloads of
+   a _local_ token are always encrypted, it is impractical to store the
+   key id in the payload.
+
+   Instead, users should store Key-ID claims (_kid_) in the unencrypted
+   footer.
+
+   For example, a footer of {"kid":"gandalf0"} can be read without
+   needing to first decrypt the token (which would in turn allow the
+   user to know which key to use to decrypt the token).
+
+   Implementations should feel free to provide a means to extract the
+   footer from a token, before authentication and decryption.  This is
+   possible for _local_ tokens because the contents of the footer are
+   _not_ encrypted.  However, the authenticity of the footer is only
+   assured after the authentication tag is verified.
+
+
+
+
+Arciszewski & Haussmann Expires October 21, 2018               [Page 16]
+
+Internet-Draft                                                April 2018
+
+
+   While a key identifier can generally be safely used for selecting the
+   cryptographic key used to decrypt and/or verify payloads before
+   verification, provided that the "key-id" is a public number that is
+   associated with a particular key which is not supplied by attackers,
+   any other fields stored in the footer MUST be distrusted until the
+   payload has been verified.
+
+   IMPORTANT: Key identifiers MUST be independent of the actual keys
+   used by Paseto.
+
+   For example, the user MUST NOT store the public key in the footer for
+   a *public* token and have the recipient use the provided public key.
+   Doing so would allow an attacker to replace the public key with one
+   of their own choosing, which will cause the recipient to accept any
+   signature for any message as valid, which defeats the security goals
+   of public-key cryptography.
+
+   Instead, it's recommended that implementors and users use a unique
+   identifier for each key (independent of the cryptographic key's
+   contents) that is used in a database or other key-value store to
+   select the appropriate cryptographic key.  These search operations
+   MUST fail closed if no valid key is found for the given key
+   identifier.
```

Comment:
Being pedantic, but a thumbprint seems like it would be a satisfactory identifier, but is not "independent" of the key itself.